source/extensions/filters/network/redis_proxy/command_splitter_impl.cc

```diff
@@ -432,13 +446,98 @@ void SplitKeysSumResultRequest::onChildResponse(Common::Redis::RespValuePtr&& va
   }
 }
 
+SplitRequestPtr TransactionRequest::create(Router& router,
+                                           Common::Redis::RespValuePtr&& incoming_request,
+                                           SplitCallbacks& callbacks, CommandStats& command_stats,
+                                           TimeSource& time_source, bool delay_command_latency) {
+  Common::Redis::Client::Transaction& transaction = callbacks.transaction();
+  std::string to_lower_string = absl::AsciiStrToLower(incoming_request->asArray()[0].asString());
+
+  // Within transactions we only support simple commands.
+  // So if this is not a transaction command or a simple command, it is an error.
+  if (Common::Redis::SupportedCommands::transactionCommands().count(to_lower_string) == 0 &&
+      Common::Redis::SupportedCommands::simpleCommands().count(to_lower_string) == 0) {
+    callbacks.onResponse(Common::Redis::Utility::makeError(
+        fmt::format("'{}' command is not supported within transaction",
+                    incoming_request->asArray()[0].asString())));
+    return nullptr;
+  }
+
+  // Start transaction on MULTI, and stop on EXEC/DISCARD.
+  if (to_lower_string == "multi") {
+    // Check for nested MULTI commands.
+    if (transaction.active_) {
+      callbacks.onResponse(
+          Common::Redis::Utility::makeError(fmt::format("MULTI calls can not be nested")));
+      return nullptr;
+    }
+    transaction.start();
+    // Respond to MULTI locally.
+    localResponse(callbacks, "OK");
+    return nullptr;
+
+  } else if (to_lower_string == "exec" || to_lower_string == "discard") {
+    // Handle the case where we don't have an open transaction.
+    if (transaction.active_ == false) {
+      callbacks.onResponse(Common::Redis::Utility::makeError(
+          fmt::format("{} without MULTI", absl::AsciiStrToUpper(to_lower_string))));
+      return nullptr;
+    }
+
+    // Handle the case where the transaction is empty.
+    if (transaction.key_.empty()) {
+      if (to_lower_string == "exec") {
+        localResponse(callbacks, "(empty array)");
+      } else {
+        localResponse(callbacks, "OK");
+      }
+      transaction.close();
+      return nullptr;
+    }
+
+    // In all other cases we will close the transaction connection after sending the last command.
+    transaction.should_close_ = true;
+  }
+
+  // When we receive the first command with a key we will set this key as our transaction
+  // key, and then send a MULTI command to the node that handles that key.
+  const auto route = router.upstreamPool(transaction.key_);
```

Comment:
When transaction.key_ is empty (for the first command), should we first assign transaction.key_ and then calculate the route?

What will happen if the keys of subsequent commands (not the first) hash to different shards? I guess the connected Redis node will return "MOVED" error and we will return the "MOVED" errors back to the client

Reply:
You are correct, fixed.